File: string/1496_path_crossing/1496_path_crossing.go
package main

import "fmt"

func isPathCrossing(path string) bool {
	visited := make(map[string]bool)

	visited["00"] = true

	x := 0
	y := 0

	for _, val := range path {
		switch val {
		case 'N':
			y++
		case 'S':
			y--
		case 'E':
			x++
		default:
			x--
		}

		key := fmt.Sprintf("%d%d", x, y)

		_, ok := visited[key]

		if ok {
			return true
		}

		visited[key] = true
	}

	return false
}

func main() {
	path := "NES"
	res := isPathCrossing(path)
	fmt.Println(res)
}

File: string/1496_path_crossing/1496_path_crossing.py
class Solution:
    def isPathCrossing(self, path: str) -> bool:
        x = y = 0
        set = {(0, 0)}
        
        for c in path:
            if c == 'N':
                y += 1
            elif c == 'S':
                y -= 1
            elif c == 'E':
                x += 1
            else:
                x -= 1
            
            if (x, y) in set:
                return True
            set.add((x, y))
        return False


path = "NES"
res = Solution().isPathCrossing(path)
print(res)